pkg/receive/handler.go

```diff
@@ -1123,46 +1161,69 @@ func newReplicationErrors(threshold, numErrors int) []*replicationErrors {
 	return errs
 }
 
-func newPeerGroup(dialOpts ...grpc.DialOption) *peerGroup {
+func newPeerGroup(dialOpts ...grpc.DialOption) peersContainer {
 	return &peerGroup{
 		dialOpts: dialOpts,
-		cache:    map[string]storepb.WriteableStoreClient{},
+		cache:    map[string]*grpc.ClientConn{},
 		m:        sync.RWMutex{},
 		dialer:   grpc.DialContext,
 	}
 }
 
+type peersContainer interface {
+	close(string) error
+	get(context.Context, string) (storepb.WriteableStoreClient, error)
+}
+
 type peerGroup struct {
 	dialOpts []grpc.DialOption
-	cache    map[string]storepb.WriteableStoreClient
+	cache    map[string]*grpc.ClientConn
 	m        sync.RWMutex
 
 	// dialer is used for testing.
 	dialer func(ctx context.Context, target string, opts ...grpc.DialOption) (conn *grpc.ClientConn, err error)
 }
 
+func (p *peerGroup) close(addr string) error {
+	p.m.Lock()
+	defer p.m.Unlock()
+
+	c, ok := p.cache[addr]
+	if !ok {
+		// NOTE(GiedriusS): this could be valid case when the connection
+		// was never established.
+		return nil
+	}
+
+	if err := c.Close(); err != nil {
+		return fmt.Errorf("closing connection for %s", addr)
+	}
+
+	delete(p.cache, addr)
```

Comment:
should we delete it from the cache before closing?

Reply:
IMHO it's probably safer to keep trying to close the connection if closing fails

Reply:
Im worrying that if an address leaves hashring and closing fails if it now rejoins we get closed connection from the peerGroup cache. ( Or connection in a bad state anyway )

Reply:
Okay, I've updated this part. Please take a look :hugs: